### PR TITLE
Add simple performance gate to integration tests

### PR DESCRIPTION
**Pull request by julienp**
Add a performance gate for PRs and merges to the master branch. The performance tests can be run locally using `make test_performance`.

This PR adds an initial batch of performance tests, all using Python:

* TestPerfEmptyUpdate: an empty Python program that does nothibg
* TestPerfManyComponentUpdate: a Python program that creates 100 resources
* TestPerfParentChainUpdate: a Python program that creates 100 resources, parented in a chain

More tests for other scenarios and languages can be added to `tests/performance/performance_test.go`. 

The tests are run in a separate GitHub actions workflow so that we can use binaries that are built without coverage instrumentation or race detection, which could otherwise have an impact on performance. This also ensures that we use the same setup in PRs and in the merge queue.

The thresholds used to determine if a test has passed or failed are highly dependent on the GitHub Actions Runners. Initial thresholds have been set by running the tests multiple times, taking the slowest run, and adding 10% (rounded to 100ms). These thresholds are not perfect and may need to be adjusted over time.

### TestPerfEmptyUpdate
5.06
5.11
4.82
4.95
5.74

### TestPerfManyComponentUpdate
16.7
17.1
16.62
16.28
16.29

### TestPerfParentChainUpdate
17.58
17.58
17.46
17.23
17.34

Fixes https://github.com/pulumi/pulumi/issues/15347

### CODE REVIEW

**Comment by julienp** [comment]
Maybe we should run these in a separate job, and sequentially, to avoid too much noise and contention with other tests.

File: tests/performance/README.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# Performance Tests
+
+This package contains basic performance tests for the Pulumi CLI. These tests are intended to run as part of pull request and prevent us from introducing major performance regressions.
+
+The [cli-performance-metrics](https://github.com/pulumi/cli-performance-metrics) repository contains more comprehensive performance tests that regularly run as a cron job.
+
+## Thresholds
+
+The thresholds used to determine if a test has passed or failed are highly dependent on the GitHub Actions Runners. Initial thresholds have been set by running the tests multiple times, taking the slowest run, and adding 10% (rounded to 100ms). These thresholds are not perfect and may need to be adjusted over time.
```
Comment (julienp) [inline tests/performance/README.md:9]: > 10%

Maybe that's not enough to ensure we don't get flakes?

File: scripts/get-job-matrix.py [hunk 2]
```diff
@@ -72,6 +76,9 @@ def is_unit_test(pkg: str) -> bool:
         or pkg in INTEGRATION_TEST_PACKAGES
     )
 
+def is_performance_test(pkg: str) -> bool:
```
Comment (tgummerer) [inline scripts/get-job-matrix.py:79]: I can't comment here, but should the return in `is_unit_test` above now be changed to

```
return not (
        pkg.startswith("github.com/pulumi/pulumi/tests")
        or pkg in INTEGRATION_TEST_PACKAGES
        or pkg in PERFORMANCE_TEST_PACKAGES
    )
```

?

File: tests/performance/README.md (hunk 1 above)
Comment (tgummerer) [reply tests/performance/README.md:9]: I'm fine with giving it a try, and adjusting as we need.

**Review by tgummerer — COMMENTED** [review]
I'm a little curious about the choice of Python for these performance tests. Are we planning to add more for NodeJS/Go?  Or are we just testing the engine here (in which case Go might be better since we don't need to spin up as much language runtime?)

File: scripts/get-job-matrix.py (hunk 2 above)
Comment (julienp) [reply scripts/get-job-matrix.py:79]: Yes, thanks!

**Comment by julienp** [comment]
> I'm a little curious about the choice of Python for these performance tests. Are we planning to add more for NodeJS/Go? Or are we just testing the engine here (in which case Go might be better since we don't need to spin up as much language runtime?)

Sorry, I missed this. There's no real reason for Python over other languages, it's just what I picked for to create the examples.

I think it could make sense to follow up with a test for each language. I suspect the most useful thing would be to come up with more complex scenarios in the programs, cases where things should be fast because they happen in parallel, but that would slow down if we started processing something serially.

File: .github/workflows/ci-build-binaries.yml [hunk 3]
```diff
@@ -57,10 +61,6 @@ jobs:
   build:
     name: ${{ inputs.os }}-${{ inputs.arch }}
     runs-on: ${{ inputs.build-platform }}
-
-    permissions:
-      id-token: write
```
Comment (julienp) [inline .github/workflows/ci-build-binaries.yml:62]: TODO: move this to a separate PR

This was added in https://github.com/pulumi/pulumi/pull/11310, but I don't *think* this is needed

**Review by tgummerer — APPROVED** [review]
Neat, let's give it a try.

File: .github/workflows/ci-performance-gate.yml [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+      version: ${{ inputs.version }}
+      os: ${{ matrix.target.os }}
+      arch: ${{ matrix.target.arch }}
+      build-platform: ${{ matrix.target.build-platform }}
+      version-set: ${{ needs.matrix.outputs.version-set }}
+      # For performance tests, we do not want coverage or race detection enabled.
+      enable-coverage: false
+      enable-race-detection: false
+      # Suffix the artifacts so they do not clash with those of the main build.
+      artifact-suffix: '-perf'
```
[the remark this answers is not in the document]
Comment (julienp) [reply .github/workflows/ci-performance-gate.yml:142]: Oh that's a great idea!
